base/collection_xattr_common.go

```diff
@@ -0,0 +1,360 @@
+	case mutateErr == ErrNotFound:
+		// Invoke the testing related callback.  This is a no-op in non-test contexts.
+		if callback != nil {
+			callback(k, xattrKey)
+		}
+		// KeyNotFound indicates there is no doc body.  Try to delete only the xattr.
+		return deleteDocXattrOnly(store, k, xattrKey, callback)
+	case mutateErr == ErrXattrNotFound:
+		// Invoke the testing related callback.  This is a no-op in non-test contexts.
+		if callback != nil {
+			callback(k, xattrKey)
+		}
+		// ErrXattrNotFound indicates there is no XATTR.  Try to delete only the body.
+		return store.Delete(k)
+
+	default:
+		// return error
+		return mutateErr
+	}
+
+}
+
+func deleteDocXattrOnly(store SubdocXattrStore, k string, xattrKey string, callback deleteWithXattrRaceInjection) error {
+
+	//  Do get w/ xattr in order to get cas
+	var retrievedVal map[string]interface{}
+	var retrievedXattr map[string]interface{}
+	getCas, err := store.SubdocGetBodyAndXattr(k, xattrKey, "", &retrievedVal, &retrievedXattr, nil)
+	if err != nil {
+		return err
+	}
+
+	// If the doc body is non-empty at this point, then give up because it seems that a doc update has been
+	// interleaved with the purge.  Return error to the caller and cancel the purge.
+	if len(retrievedVal) != 0 {
+		return fmt.Errorf("DeleteWithXattr was unable to delete the doc. Another update " +
+			"was received which resurrected the doc by adding a new revision, in which case this delete operation is " +
+			"considered as cancelled.")
+	}
+
+	// Cas-safe delete of just the XATTR.  Use SubdocDocFlagAccessDeleted since presumably the document body
+	// has been deleted.
+	deleteXattrErr := store.SubdocDeleteXattr(k, xattrKey, getCas)
+	if deleteXattrErr != nil {
+		// If the cas-safe delete of XATTR fails, return an error to the caller.
+		// This might happen if there was a concurrent update interleaved with the purge (someone resurrected doc)
+		return pkgerrors.Wrapf(deleteXattrErr, "DeleteWithXattr was unable to delete the doc.  Another update "+
+			"was received which resurrected the doc by adding a new revision, in which case this delete operation is "+
+			"considered as cancelled. ")
+	}
+	return nil
+
+}
+
+// AsSubdocXattrStore tries to return the given bucket as a SubdocXattrStore, based on underlying buckets.
+func AsSubdocXattrStore(bucket Bucket) (SubdocXattrStore, bool) {
+
+	var underlyingBucket Bucket
+	switch typedBucket := bucket.(type) {
+	case *CouchbaseBucketGoCB:
+		return typedBucket, true
+	case *Collection:
+		return typedBucket, true
+	case *LoggingBucket:
+		underlyingBucket = typedBucket.GetUnderlyingBucket()
+	case *LeakyBucket:
+		underlyingBucket = typedBucket.GetUnderlyingBucket()
+	case *TestBucket:
+		underlyingBucket = typedBucket.Bucket
+	default:
+		// bail out for unrecognised/unsupported buckets
+		return nil, false
+	}
+
+	return AsSubdocXattrStore(underlyingBucket)
+}
+
+func xattrCasPath(xattrKey string) string {
+	return fmt.Sprintf("%s.%s", xattrKey, xattrMacroCas)
```

Comment:
String concatenation is quite a bit faster and lower/zero alloc vs. Sprintf. Given this is in the doc-write path, it's probably worth doing.

```
BenchmarkStrConcat-12    	73450123	        16.38 ns/op	       0 B/op	       0 allocs/op
BenchmarkSprintf-12      	 9484422	       125.2 ns/op	      24 B/op	       2 allocs/op
```

```suggestion
	return xattrKey + "." + xattrMacroCas
```